sql/2025-04-30-contribution-diff-queue.sql

```diff
@@ -0,0 +1,13 @@
+-- Drop and re-create the contribution_diff_queue table without a primary key constraint on contribution_id.
+-- This allows us to insert redundant rows while a diff is being computed, rather than blocking due to row lock.
```

Comment:
I think I'm missing the "why" of why we want redundant rows while the diff is being computed? If it's already in the queue should we perhaps just `ON CONFLICT DO NOTHING`?

Reply:
Whoops, I'm realizing that the disadvantage of this in-between design is that we can enqueue one job multiple times that gets picked up and worked on in parallel, which we don't want.

Ok, I think the two reasonable options, then, are:

- Keep existing design for job queue, but keep uniqueness constraint, and remove line in "get contribution diff" endpoint that tries to enqueue a job if the diff hasn't been computed.
- Move to a design that involves claiming work explicitly in the database, with a recorded timestamp, and infer that a job that began X minutes ago but hasn't been completed is safe to start again.

I sorta lean towards the first option, for simplicity.

Reply:
Yeah, I'm going to go with the first. The second is more robust but would require building a little heartbeating mechanism, and feels like overkill for now.

Reply:
The first one sounds good, however what happens for all the old existing diffs? If we don't enqueue anything when you load the page won't those diffs simply never be computed?

I'm no expert on PG row locking, is there a Transaction Isolation level where you could run a quick query to see if the row already exists, then proceed with an insert with "ON CONFLICT DO NOTHING" in a follow-up transaction if it appears it doesn't already exist?